src/main/java/org/traccar/handler/ComputedAttributesHandler.java

```diff
@@ -102,7 +115,7 @@ private MapContext prepareContext(Position position) {
      */
     @Deprecated
     public Object computeAttribute(Attribute attribute, Position position) throws JexlException {
-        return engine.createExpression(attribute.getExpression()).evaluate(prepareContext(position));
+        return engine.createScript(features, engine.createInfo(), attribute.getExpression()).execute(prepareContext(position));
```

Comment:
What is the difference between `createExpression().evaluate()` and `createScript().execute()`?

Reply:
Let's start with the difference between expressions and scripts: [docs say](https://commons.apache.org/proper/commons-jexl/apidocs/org/apache/commons/jexl3/JexlExpression.html) that expression is a single, simple expression, and no declarations, loops or blocks are allowed there. Scripts, however, are a different story - they do exactly the same things `JexlExpression` does, but also allows for local variables, loops and so on to be defined and used. And, after all, I had to use `createScript()` to plug the `JexlFeatures` object during the expression execution.

`.execute()` and `.evaluate()` do the same thing (execute the given expression), they're just named different for some reason.

Reply:
A follow up question. With variables, are they persisted between executions or they are scoped to a single execution and cleared after?

Reply:
They're in their own scope. When the execution is compete, the variables are released.